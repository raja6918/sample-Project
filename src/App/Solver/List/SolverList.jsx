import React, { Component, Fragment, createRef } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import PerfectScrollbar from 'react-perfect-scrollbar';
import 'react-perfect-scrollbar/dist/css/styles.css';

import IconButton from '@material-ui/core/IconButton';
import Input from '@material-ui/core/Input';
import Collapse from '@material-ui/core/Collapse';
import CheckboxMUI from '@material-ui/core/Checkbox';
import Menu from '@material-ui/core/Menu';

import MenuItem from '../../../components/Menu/MenuItem';
import MenuText from '../../../components/Menu/MenuText';
import MenuAction from '../../../components/Menu/MenuAction';
import Icon from '../../../components/Icon';

import rocket from './rocket.svg';
import { perfectScrollConfig } from '../../../utils/common';

import {
  InputSearch,
  SearchIcon,
  ClearButton,
  ClearIcon,
} from '../../../components/FilterInput';

import SolverCard from './SolverCard';
import AccessEnabler from '../../../components/AccessEnabler';

const Checkbox = styled(CheckboxMUI)`
  font-size: 1.5rem;
  vertical-align: top;
  width: 20px;
  /* padding: 12px; */
`;

const List = styled.ul`
  height: calc(100% - 48px);
  margin: 0;
  padding: 0;
  /* overflow-y: auto;
  overflow-x: hidden; */
  border-right: 1px solid #cecece;
`;

const FilterContainer = styled.div`
  height: 48px;
  border-bottom: 1px solid #cecece;
  background-color: #fff;
  padding: 12px;
  & button:last-child {
    float: right;
    z-index: 1000;
  }
  & .MuiIconButton-root:hover {
    /* background-color: transparent !important; */
  }
`;

const FilterInputContainer = styled(Collapse)`
  text-align: center;
  padding: ${props => (props.in ? '5px' : '0')} 0;
  border-bottom: ${props => (props.in ? '1px solid #cecece' : 'none')};
  background-color: #fff;
`;

class SolverList extends Component {
  state = {
    requests: this.props.solverRequests,
    filteredRequests: this.props.solverRequests,
    showFilter: false,
    action: this.props.action,
    multiSelected: [],
    anchorEl: null,
  };
  searchTerm = '';

  filter = createRef();

  componentDidMount() {
    window.addEventListener('resize', this.shakeScrollBar);
  }

  componentWillReceiveProps(nextProps) {
    const updatedMultiselected = [];
    const { multiSelected, showFilter } = this.state;
    if (multiSelected.length) {
      multiSelected.forEach(selected => {
        let i = 0;
        while (i < nextProps.solverRequests.length) {
          if (selected.id === nextProps.solverRequests[i].id) {
            updatedMultiselected.push(nextProps.solverRequests[i]);
            break;
          }
          i++;
        }
      });
    }

    this.setState({
      requests: nextProps.solverRequests,
      action: nextProps.action,
      filteredRequests: showFilter
        ? this.filterRequests(nextProps.solverRequests)
        : nextProps.solverRequests,
      multiSelected: updatedMultiselected,
    });
  }

  toggleFilter = () => {
    this.setState(state => {
      if (state.showFilter) {
        this.filter.current.value = '';
        return { filteredRequests: this.state.requests, showFilter: false };
      } else {
        return { showFilter: true };
      }
    });
  };

  isEmpty = () =>
    this.filter.current ? this.filter.current.value.length < 1 : true;

  onFilter = () => {
    const { requests } = this.state;
    this.setState({ filteredRequests: this.filterRequests(requests) });
  };

  filterRequests = requests => {
    if (this.filter.current.value !== '')
      return requests.filter(
        val =>
          val.name
            .toLowerCase()
            .indexOf(this.filter.current.value.toLowerCase()) !== -1
      );
    return this.state.requests;
  };

  clearFilter = () => {
    this.filter.current.value = '';
    this.setState({ filteredRequests: this.state.requests });
  };

  selectAllClick = (event, checked) => {
    if (checked) {
      const validSolversToCompare = this.state.filteredRequests.filter(
        solver => {
          return solver.status.status === 'Done-success';
        }
      );
      this.setState({ multiSelected: this.state.filteredRequests }, () => {
        if (validSolversToCompare.length > 25) {
          this.props.handleCompareAll([...validSolversToCompare.slice(0, 25)]);
        } else {
          this.props.handleCompareAll(validSolversToCompare);
        }
      });
    } else {
      this.setState({ multiSelected: [] }, () => {
        this.props.handleCompare([]);
      });
    }
  };

  isSelected = id =>
    this.state.multiSelected.find(val => val.id === id) !== undefined;

  handleCheckbox = (event, request) => {
    const { multiSelected } = this.state;
    const selectedRequest = multiSelected.find(val => val.id === request.id);
    let newMultiSelected = [];
    let action = 'add';

    if (selectedRequest) {
      newMultiSelected = multiSelected.filter(
        val => val.id !== selectedRequest.id
      );
      action = 'remove';
    } else {
      newMultiSelected = [...multiSelected, request];
    }
    this.setState({ multiSelected: newMultiSelected }, () =>
      this.props.handleSolverToCompare(request, action)
    );
  };

  openMenu = event => {
    this.setState({ anchorEl: event.currentTarget });
  };

  closeMenu = () => {
    this.setState({ anchorEl: null });
  };

  handleCompare = () => {
    if (this.state.multiSelected.length > 0) {
      const validSolverIds = this.state.multiSelected
        .filter(solver => {
          return solver.status.status === 'Done-success';
        })
        .map(request => request.id);

      const validSolversToCompare = this.state.requests.filter(solver => {
        return validSolverIds.includes(solver.id);
      });

      if (validSolversToCompare.length > 10) {
        this.props.handleCompare([...validSolversToCompare.slice(0, 10)]);
      } else {
        this.props.handleCompare(validSolversToCompare);
      }

      this.closeMenu();
    }
  };

  isReadyToCompare = () => {
    const { multiSelected } = this.state;
    let isDisabled = true;
    let i = 0;
    for (const selected of multiSelected) {
      if (selected.status.status === 'Done-success') {
        i++;
      }
      if (i >= 1) {
        isDisabled = false;
        break;
      }
    }
    return isDisabled;
  };

  shakeScrollBar = () => {
    // shake the scroll to get the scrollbar active in case of resizes :(
    setTimeout(() => {
      if (this._scrollRefY) {
        this._scrollRefY.scrollTop += 1;
      }
    }, 1000);
  };

  render() {
    const { active, handleClick, t, readOnly, solverScopes } = this.props;
    const {
      showFilter,
      filteredRequests,
      action,
      multiSelected,
      anchorEl,
    } = this.state;

    const numSelected = multiSelected.length;
    const rowCount = filteredRequests.length;
    const isDisabled =
      multiSelected.filter(solver => solver.status.status === 'Done-success')
        .length < 1;

    return (
      <Fragment>
        <FilterContainer open={showFilter}>
          <Checkbox
            onChange={this.selectAllClick}
            indeterminate={numSelected > 0 && numSelected < rowCount}
            checked={numSelected === rowCount && rowCount > 0}
            style={{ paddingRight: '5px', marginRight: '15px' }}
          />
          <IconButton onClick={this.toggleFilter}>
            <Icon iconcolor="#0A75C2" margin="0">
              filter_list
            </Icon>
          </IconButton>
          <IconButton
            aria-label="More"
            aria-owns={anchorEl ? 'long-menu' : null}
            aria-haspopup="true"
            onClick={this.openMenu}
          >
            <Icon iconcolor="#0A75C2" margin="0">
              more_vert
            </Icon>
          </IconButton>
          <Menu
            id="long-menu"
            anchorEl={anchorEl}
            open={Boolean(anchorEl)}
            onClose={this.closeMenu}
          >
            <AccessEnabler
              scopes={solverScopes.solverEvaluate}
              disableComponent
              render={props => (
                <MenuAction
                  disabled={this.isReadyToCompare() || props.disableComponent}
                  handleClick={this.handleCompare}
                  icon={'compare'}
                  text={t('SOLVER.options.compare')}
                />
              )}
            />

            {/* <MenuItem disabled={isDisabled || readOnly} onClick={() => {}}>
              <img
                style={{ marginRight: 8, width: 24, height: 24 }}
                src={rocket}
                alt={t('SOLVER.options.launch')}
              />
              <MenuText>{t('SOLVER.options.launch')}</MenuText>
            </MenuItem> */}
            {/* <MenuAction
              disabled={isDisabled || readOnly}
              handleClick={() => {}}
              icon={'delete'}
              text={t('SOLVER.options.delete')}
            /> */}
          </Menu>
        </FilterContainer>
        <List>
          <PerfectScrollbar
            option={perfectScrollConfig}
            containerRef={ref => {
              this._scrollRefY = ref;
            }}
          >
            <div>
              <FilterInputContainer in={showFilter}>
                <InputSearch>
                  <Input
                    type="text"
                    disableUnderline={true}
                    startAdornment={<SearchIcon>search</SearchIcon>}
                    placeholder={t('SEARCHENGINE')}
                    onChange={e => {
                      this.onFilter(e.target.value);
                    }}
                    inputRef={this.filter}
                    endAdornment={
                      !this.isEmpty() && (
                        <ClearButton onClick={this.clearFilter}>
                          <ClearIcon>highlight_off</ClearIcon>
                        </ClearButton>
                      )
                    }
                  />
                </InputSearch>
              </FilterInputContainer>
              {filteredRequests.map(request => (
                <SolverCard
                  key={request.id}
                  request={request}
                  active={request.id === active}
                  handleClick={handleClick}
                  action={action}
                  t={t}
                  handleCheckbox={this.handleCheckbox}
                  isSelected={this.isSelected(request.id)}
                />
              ))}
            </div>
          </PerfectScrollbar>
        </List>
      </Fragment>
    );
  }
}

SolverList.propTypes = {
  solverRequests: PropTypes.arrayOf(
    PropTypes.shape({
      crewGroupName: PropTypes.string.isRequired,
      elapsedTime: PropTypes.string,
      id: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
      isEndorsed: PropTypes.bool.isRequired,
      name: PropTypes.string.isRequired,
      status: PropTypes.object.isRequired,
    })
  ).isRequired,
  active: PropTypes.number.isRequired,
  handleClick: PropTypes.func.isRequired,
  action: PropTypes.string,
  handleCompare: PropTypes.func.isRequired,
  handleSolverToCompare: PropTypes.func.isRequired,
  t: PropTypes.func.isRequired,
  handleCompareAll: PropTypes.func.isRequired,
  readOnly: PropTypes.bool,
  solverScopes: PropTypes.shape([]).isRequired,
};

SolverList.defaultProps = {
  action: '',
  readOnly: false,
};

export default SolverList;
