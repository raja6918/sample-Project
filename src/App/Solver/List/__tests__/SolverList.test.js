import React from 'react';
import { shallow } from 'enzyme';

import SolverList from '../SolverList';

const t = jest.fn();
const handleClick = jest.fn();

const solverRequests = [
  {
    id: 1,
    name: 'Crew Solution: 777 Pilots',
    status: 'running',
    lastModified: '2018-04-11T22:25:09.586Z'
  },
  {
    id: 2,
    name: 'Crew Solution: 888 Pilots',
    status: 'failed',
    lastModified: '2018-04-14T22:25:09.586Z'
  },
  {
    id: 3,
    name: 'Crew Solution: 999 Pilots',
    status: 'completed',
    lastModified: '2018-04-01T22:25:09.586Z'
  },
  {
    id: 4,
    name: 'Crew Solution: 777 Pilots',
    status: 'stopping',
    lastModified: '2018-04-15T22:25:09.586Z'
  },
  {
    id: 5,
    name: 'Crew Solution: 333 Pilots',
    status: 'stoppedWithResults',
    lastModified: '2018-06-20T22:25:09.586Z'
  },
  {
    id: 6,
    name: 'Crew Solution: 444 Pilots',
    status: 'completed',
    lastModified: '2018-02-11T22:25:09.586Z'
  },
  {
    id: 7,
    name: 'Crew Solution: 111 Pilots',
    status: 'running',
    lastModified: '2018-03-11T22:25:09.586Z'
  },
  {
    id: 8,
    name: 'Crew Solution: 777 Pilots',
    status: 'completed',
    lastModified: '2018-05-11T22:25:09.586Z'
  },
  {
    id: 9,
    name: 'Crew Solution: 555 Pilots',
    status: 'running',
    lastModified: '2018-01-11T22:25:09.586Z'
  },
  {
    id: 10,
    name: 'Crew Solution: 888 Pilots',
    status: 'completed',
    lastModified: '2018-09-11T22:25:09.586Z'
  },
  {
    id: 11,
    name: 'Crew Solution: 000 Pilots',
    status: 'running',
    lastModified: '2017-04-11T22:25:09.586Z'
  }
];

test.skip('SolverList Render', () => {
  expect(
    shallow(
      <SolverList
        solverRequests={solverRequests}
        active={-1}
        handleClick={handleClick}
        t={t}
      />
    )
  ).toMatchSnapshot();
});
