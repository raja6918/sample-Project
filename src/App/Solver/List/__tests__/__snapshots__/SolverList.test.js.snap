// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SolverList Render 1`] = `
<React.Fragment>
  <SolverList__FilterContainer
    open={false}
  >
    <SolverList__Checkbox
      checked={false}
      indeterminate={false}
      onChange={[Function]}
    />
    <WithStyles(IconButton)
      onClick={[Function]}
    >
      <Icon
        iconcolor="#0A75C2"
        margin="0"
      >
        filter_list
      </Icon>
    </WithStyles(IconButton)>
    <WithStyles(IconButton)
      aria-haspopup="true"
      aria-label="More"
    >
      <Icon
        iconcolor="#0A75C2"
        margin="0"
      >
        more_vert
      </Icon>
    </WithStyles(IconButton)>
  </SolverList__FilterContainer>
  <SolverList__List>
    <SolverList__FilterInputContainer
      in={false}
    >
      <FilterInput__InputSearch>
        <WithStyles(Input)
          disableUnderline={true}
          endAdornment={false}
          innerRef={[Function]}
          onChange={[Function]}
          placeholder="search"
          startAdornment={
            <FilterInput__SearchIcon>
              search
            </FilterInput__SearchIcon>
          }
          type="text"
        />
      </FilterInput__InputSearch>
    </SolverList__FilterInputContainer>
    <SolverList__Item
      key="6"
    >
      <SolverCard
        active={false}
        handleCheckbox={[Function]}
        handleClick={[MockFunction]}
        isSelected={false}
        request={
          Object {
            "id": 6,
            "lastModified": "2018-02-11T22:25:09.586Z",
            "name": "Crew Solution: 444 Pilots",
            "status": "completed",
          }
        }
        t={[MockFunction]}
      />
    </SolverList__Item>
    <SolverList__Item
      key="1"
    >
      <SolverCard
        active={false}
        handleCheckbox={[Function]}
        handleClick={[MockFunction]}
        isSelected={false}
        request={
          Object {
            "id": 1,
            "lastModified": "2018-04-11T22:25:09.586Z",
            "name": "Crew Solution: 777 Pilots",
            "status": "running",
          }
        }
        t={[MockFunction]}
      />
    </SolverList__Item>
    <SolverList__Item
      key="3"
    >
      <SolverCard
        active={false}
        handleCheckbox={[Function]}
        handleClick={[MockFunction]}
        isSelected={false}
        request={
          Object {
            "id": 3,
            "lastModified": "2018-04-01T22:25:09.586Z",
            "name": "Crew Solution: 999 Pilots",
            "status": "completed",
          }
        }
        t={[MockFunction]}
      />
    </SolverList__Item>
    <SolverList__Item
      key="4"
    >
      <SolverCard
        active={false}
        handleCheckbox={[Function]}
        handleClick={[MockFunction]}
        isSelected={false}
        request={
          Object {
            "id": 4,
            "lastModified": "2018-04-15T22:25:09.586Z",
            "name": "Crew Solution: 777 Pilots",
            "status": "stopping",
          }
        }
        t={[MockFunction]}
      />
    </SolverList__Item>
    <SolverList__Item
      key="5"
    >
      <SolverCard
        active={false}
        handleCheckbox={[Function]}
        handleClick={[MockFunction]}
        isSelected={false}
        request={
          Object {
            "id": 5,
            "lastModified": "2018-06-20T22:25:09.586Z",
            "name": "Crew Solution: 333 Pilots",
            "status": "stoppedWithResults",
          }
        }
        t={[MockFunction]}
      />
    </SolverList__Item>
    <SolverList__Item
      key="2"
    >
      <SolverCard
        active={false}
        handleCheckbox={[Function]}
        handleClick={[MockFunction]}
        isSelected={false}
        request={
          Object {
            "id": 2,
            "lastModified": "2018-04-14T22:25:09.586Z",
            "name": "Crew Solution: 888 Pilots",
            "status": "failed",
          }
        }
        t={[MockFunction]}
      />
    </SolverList__Item>
    <SolverList__Item
      key="7"
    >
      <SolverCard
        active={false}
        handleCheckbox={[Function]}
        handleClick={[MockFunction]}
        isSelected={false}
        request={
          Object {
            "id": 7,
            "lastModified": "2018-03-11T22:25:09.586Z",
            "name": "Crew Solution: 111 Pilots",
            "status": "running",
          }
        }
        t={[MockFunction]}
      />
    </SolverList__Item>
    <SolverList__Item
      key="8"
    >
      <SolverCard
        active={false}
        handleCheckbox={[Function]}
        handleClick={[MockFunction]}
        isSelected={false}
        request={
          Object {
            "id": 8,
            "lastModified": "2018-05-11T22:25:09.586Z",
            "name": "Crew Solution: 777 Pilots",
            "status": "completed",
          }
        }
        t={[MockFunction]}
      />
    </SolverList__Item>
    <SolverList__Item
      key="9"
    >
      <SolverCard
        active={false}
        handleCheckbox={[Function]}
        handleClick={[MockFunction]}
        isSelected={false}
        request={
          Object {
            "id": 9,
            "lastModified": "2018-01-11T22:25:09.586Z",
            "name": "Crew Solution: 555 Pilots",
            "status": "running",
          }
        }
        t={[MockFunction]}
      />
    </SolverList__Item>
    <SolverList__Item
      key="10"
    >
      <SolverCard
        active={false}
        handleCheckbox={[Function]}
        handleClick={[MockFunction]}
        isSelected={false}
        request={
          Object {
            "id": 10,
            "lastModified": "2018-09-11T22:25:09.586Z",
            "name": "Crew Solution: 888 Pilots",
            "status": "completed",
          }
        }
        t={[MockFunction]}
      />
    </SolverList__Item>
    <SolverList__Item
      key="11"
    >
      <SolverCard
        active={false}
        handleCheckbox={[Function]}
        handleClick={[MockFunction]}
        isSelected={false}
        request={
          Object {
            "id": 11,
            "lastModified": "2017-04-11T22:25:09.586Z",
            "name": "Crew Solution: 000 Pilots",
            "status": "running",
          }
        }
        t={[MockFunction]}
      />
    </SolverList__Item>
  </SolverList__List>
</React.Fragment>
`;
